metpy/calc/tools.py

```diff
@@ -286,47 +287,137 @@ def reduce_point_density(points, radius, priority=None):
+
+    axis : int, optional
+        The axis to interpolate over. Defaults to 0.
+
+    extrap: string, optional
+        Specify handling of interpolation points out of data bounds
+        'set_nan' : default, set value of points to 'nan'.
+        'clamp'   : set values to endpoint of data range.
+        'none'    : return ValueError if data is out of bounds.
 
     Returns
     -------
     array-like
-        The interpolated values, same shape as x.
+        List of interpolated values for each point.
 
-    Examples
-    --------
-    >>> x_log = np.array([1e3, 1e4, 1e5, 1e6])
-    >>> y_log = np.log(x_log) * 2 + 3
-    >>> x_interp = np.array([5e3, 5e4, 5e5])
-    >>> metpy.calc.log_interp(x_interp, x_log, y_log)
-    array([ 20.03438638,  24.63955657,  29.24472675])
+    Notes
+    -----
+    xp and args must be the same shape.
 
     """
-    sort_args = np.argsort(xp)
+    # Pull out keyword args
+    extrap = kwargs.pop('extrap', 'set_nan')
+    axis = kwargs.pop('axis', 0)
 
+    # Remove units from x coords
     if hasattr(x, 'units'):
         x = x.m
 
     if hasattr(xp, 'units'):
         xp = xp.m
 
-    interpolated_vals = np.interp(np.log(x), np.log(xp[sort_args]), fp[sort_args], **kwargs)
+    # Save number of dimensions in xp
+    ndim = xp.ndim
 
-    if hasattr(fp, 'units'):
-        interpolated_vals = interpolated_vals * fp.units
+    # Sort input data
+    sort_args = np.argsort(xp, axis=axis)
+    sort_x = np.argsort(x)
 
-    return interpolated_vals
+    # indicies for sorting
+    sorter = broadcast_indicies(xp, sort_args, ndim, axis)
+
+    # Ensure pressure in increasing order
+    variables = []
+    for arr in args:
+        variables.append(arr[sorter])
+
+    # Make x the same shape as xp
+    if ndim == 1:
+        x_array = x[sort_x]
+    else:
+        x_array = x[sort_x]
+        for i in range(0, ndim - axis - 1):
+            x_array = x_array[..., np.newaxis]
+        shp = list(xp.shape)
+        shp[axis] = x.shape[0]
+        x_array = np.broadcast_to(x_array, shp)
+
+    # Log x and xp
+    log_x_array = np.log(x_array)
+    log_xp = np.log(xp[sorter])
+
+    # Calculate value above interpolated value
+    minv = np.apply_along_axis(np.searchsorted, axis, xp[sorter], x[sort_x])
+    minv2 = np.copy(minv)
+
+    # If extrap is none and data is out of bounds, raise value error
+    if ((np.max(minv) == xp.shape[axis]) or (np.min(minv) == 0)) and extrap == 'none':
+        raise ValueError('Interpolation point out of data bounds encountered')
+
+    # Warn if interpolated values are outside data bounds, will make these the values
+    # at end of data range.
+    if np.max(minv) == xp.shape[axis]:
+        warnings.warn('Interpolation point out of data bounds encountered')
+        minv2[minv == xp.shape[axis]] = xp.shape[axis] - 1
+    if np.max(minv) == 0:
+        warnings.warn('Interpolation point out of data bounds encountered')
+        minv2[minv == 0] = 1
+
+    # Create empty output list
```

Comment:
Stylistically, can you create the empty list as the last step right before the loop that appends to it? It's just easier to reason about the code that way.